tests_and_analysis/test/euphonic_test/test_brille_interpolator.py

```diff
@@ -0,0 +1,294 @@
+        mock_grid = mocker.patch(expected_grid_type)
+        mock_grid().rlu = np.ones((5, 3))  # Note mock_grid() parentheses
+
+        fc = get_fc('quartz')
+        BrilleInterpolator.from_force_constants(fc, **kwargs)
+        for called_key, called_val in mock_grid.call_args[1].items():
+            expected_val = expected_grid_kwargs.pop(called_key)
+            assert expected_val == pytest.approx(called_val)
+        # Make sure all expected kwargs have been passed
+        assert not expected_grid_kwargs
+
+    @pytest.mark.parametrize(
+        'kwargs', [
+            ({}),
+            ({'grid_type': 'nest',
+              'grid_kwargs': {'number_density': 50},
+              'interpolation_kwargs': {'insert_gamma': True}}),
+            ({'interpolation_kwargs': {'reduce_qpts': True,
+                                       'dipole_parameter': 0.5,
+                                       'n_threads': 2}})
+        ])
+    def test_from_force_constants_correct_interpolation_kwargs_passed(
+            self, mocker, kwargs):
+        # Patch __init__ to avoid type checks - we're mocking
+        # calculate_qpoint_phonon_modes so don't care about the return value
+        mocker.patch.object(BrilleInterpolator, '__init__', return_value=None)
+        mock_qpm = mocker.patch.object(
+            ForceConstants, 'calculate_qpoint_phonon_modes',
+            return_value=get_qpt_ph_modes_from_json(
+                'quartz', 'quartz_reciprocal_qpoint_phonon_modes.json'))
+
+        fc = get_fc('quartz')
+        BrilleInterpolator.from_force_constants(fc, **kwargs)
+
+        # default_kwargs should be set no matter what the user provides,
+        # as the input and output qpts of calculate_qpoint_phonon_modes
+        # must be the same
+        default_kwargs = {'insert_gamma': False, 'reduce_qpts': False}
+        expected_interp_kwargs = {**kwargs.get('interpolation_kwargs', {}),
+                                  **default_kwargs}
+        assert expected_interp_kwargs == mock_qpm.call_args[1]
+
+    def test_from_fc_with_invalid_grid_type_raises_value_error(self):
+        with pytest.raises(ValueError):
+            BrilleInterpolator.from_force_constants(
+                get_fc('quartz'), grid_type='unknown')
+
+    @pytest.mark.parametrize('faulty_crystal, grid_file, expected_exception',
+        [(get_fc('quartz'), 'lzo_trellis_10.hdf5', TypeError)])
+    def test_faulty_crystal_object_creation(
+            self, faulty_crystal, grid_file, expected_exception):
+        grid = get_brille_grid(grid_file)
+        with pytest.raises(expected_exception):
+            BrilleInterpolator(faulty_crystal, grid)
+
+    # Deliberately create a grid incompatible with the quartz crystal
+    # and check that it raises an error
+    @pytest.mark.parametrize('expected_exception, grid_file',
+        [(ValueError, 'lzo_trellis_10.hdf5'),
+         (ValueError, 'quartz_mesh_10_unfilled.hdf5')])
```

Comment:
So is the second case the error is raised because the grid is not filled, rather than because it is not the right crystal?

Reply:
Yes